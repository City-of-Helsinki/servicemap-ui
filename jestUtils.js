import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import english from './src/i18n/en';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: english,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });
