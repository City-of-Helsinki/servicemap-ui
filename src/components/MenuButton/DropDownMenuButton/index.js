import DropDownMenuButton from './DropDownMenuButton';

export default DropDownMenuButton;
