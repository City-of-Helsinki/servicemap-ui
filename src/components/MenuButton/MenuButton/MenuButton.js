import React from 'react';
import PropTypes from 'prop-types';
import {
  Button, ButtonBase, ClickAwayListener, Container, Divider, Typography,
} from '@mui/material';
import { FormattedMessage } from 'react-intl';
import styled from '@emotion/styled';
import { keyboardHandler } from '../../../utils';
import config from '../../../../config';

class MenuButton extends React.Component {
  state = {
    open: false,
  };

  handleToggle = () => {
    this.setState(state => ({ open: !state.open }));
  };

  handleClose = (event, refocus = false) => {
    this.setState({ open: false });
    // If refocus set to true focus back to MenuButton
    if (refocus && this.anchorEl) {
      this.anchorEl.focus();
    }
  };

  handleItemClick = (event, item) => {
    this.handleClose(event);
    item.onClick();
  };

  // Menu should close if user leaves the selection area
  closeMenuOnFocusExit = (event) => {
    const { menuItems } = this.props;
    const menuItemIds = menuItems.map(v => v.id);

    const relatedTargetId = event?.relatedTarget.id || '';
    // TODO not good to have "-map-type-radio" here
    if (!menuItemIds.includes(relatedTargetId) && !relatedTargetId.includes('-map-type-radio')) {
      this.handleClose(event);
    }
  };

  renderMenu = () => {
    const {
      panelID, children, menuHeader, menuItems, menuAriaLabel,
    } = this.props;
    return (
      <ClickAwayListener onClickAway={this.handleClose}>
        <StyledMenuPanel
          id={panelID}
          aria-label={menuAriaLabel}
          role="region"
          disableGutters
          sx={{ overflowY: panelID === 'SettingsMenuPanel' ? 'visible' : 'auto' }}
        >
          <Typography sx={{
            textAlign: 'left', fontWeight: 700, fontSize: '1.03rem', pb: 1,
          }}
          >
            <FormattedMessage id={menuHeader} />
          </Typography>
          {
            menuItems.map((v, i) => (
              <React.Fragment key={v.key}>
                <StyledMenuItemButton
                  id={v.id}
                  key={v.key}
                  role="link"
                  onClick={e => this.handleItemClick(e, v)}
                  onKeyDown={keyboardHandler(e => this.handleClose(e, true), ['esc'])}
                  onKeyPress={keyboardHandler(this.handleItemClick, ['space', 'enter'])}
                  onBlur={this.closeMenuOnFocusExit}
                  component="span"
                  tabIndex={0}
                  aria-hidden={v.ariaHidden}
                >
                  <span>{v.icon}</span>
                  <Typography sx={{ pl: 3, fontWeight: 700 }} variant="subtitle1">{v.text}</Typography>
                </StyledMenuItemButton>
                {i !== menuItems.length - 1
                  ? <Divider aria-hidden />
                  : null
                }
              </React.Fragment>
            ))
          }
          {children}
          <div aria-hidden role="button" tabIndex="0" onFocus={() => this.handleClose()} />
        </StyledMenuPanel>
      </ClickAwayListener>
    );
  };

  render() {
    const {
      buttonIcon, buttonText, id, panelID, dataSm,
    } = this.props;
    const { open } = this.state;

    return (
      <StyledContainer>
        <StyledButton
          id={id}
          data-sm={dataSm}
          ref={(node) => {
            this.anchorEl = node;
          }}
          aria-controls={open ? panelID : undefined}
          aria-haspopup="true"
          aria-expanded={open}
          onClick={this.handleToggle}
          onKeyDown={(e) => {
            if (open) {
              keyboardHandler(this.handleClose, ['esc'])(e);
            }
          }}
        >
          {
            buttonIcon && (
              <StyledIcon>{buttonIcon}</StyledIcon>
            )
          }
          <Typography component="p" variant="subtitle1">{buttonText}</Typography>
        </StyledButton>
        {
          open && this.renderMenu()
        }
      </StyledContainer>
    );
  }
}

const StyledMenuPanel = styled(Container)(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  position: 'absolute',
  maxHeight: `calc(100vh - ${config.topBarHeight}px)`,
  width: 450,
  padding: theme.spacing(2),
  paddingTop: theme.spacing(3),
  boxSizing: 'border-box',
  backgroundColor: 'white',
  color: 'black',
  zIndex: 2,
  border: `${theme.palette.detail.alpha} solid 0.5px`,
  borderRadius: 4,
  right: 0,
}));

const StyledMenuItemButton = styled(ButtonBase)(({ theme }) => ({
  height: 56,
  padding: theme.spacing(1),
  paddingRight: theme.spacing(2),
  justifyContent: 'start',
  alignItems: 'center',
  cursor: 'pointer',
  ...theme.typography.body2,
  // Text element
  '& p': {
    textAlign: 'left',
    textTransform: 'none',
    fontWeight: 'normal',
  },
  '&:hover': {
    backgroundColor: 'rgba(0, 0, 0, 0.08)',
  },
}));

const StyledContainer = styled('div')(() => ({
  alignItems: 'center',
  display: 'block',
  height: '100%',
  width: 180,
  flex: '0 1 auto',
}));

const StyledButton = styled(Button)(() => ({
  height: '100%',
  width: '100%',
  '& p': {
    textTransform: 'none',
  },
  color: 'black',
  justifyContent: 'flex-start',
}));

const StyledIcon = styled('span')(({ theme }) => ({
  color: theme.palette.primary.main,
  paddingRight: theme.spacing(1),
}));

MenuButton.propTypes = {
  buttonIcon: PropTypes.node,
  buttonText: PropTypes.string.isRequired,
  id: PropTypes.string,
  menuAriaLabel: PropTypes.string.isRequired,
  panelID: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  menuHeader: PropTypes.string.isRequired,
  dataSm: PropTypes.string.isRequired,
  menuItems: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string.isRequired,
    id: PropTypes.string,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
  })).isRequired,
};

MenuButton.defaultProps = {
  buttonIcon: null,
  id: null,
};

export default MenuButton;
