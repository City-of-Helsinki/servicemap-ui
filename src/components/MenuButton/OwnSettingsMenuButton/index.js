import OwnSettingsMenuButton from './OwnSettingsMenuButton';

export default OwnSettingsMenuButton;
