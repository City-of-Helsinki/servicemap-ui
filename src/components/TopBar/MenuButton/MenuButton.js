import { Button, Typography } from '@mui/material';
import PropTypes from 'prop-types';
import { Close, Menu } from '@mui/icons-material';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';


const MenuButton = ({
  classes, drawerOpen, pageType, toggleDrawerMenu,
}) => {
  const intl = useIntl();

  return (
    <Button
      id="MenuButton"
      aria-label={intl.formatMessage({ id: 'general.menu' })}
      aria-expanded={drawerOpen}
      aria-haspopup="true"
      className={`${drawerOpen ? classes.toolbarButtonPressed : classes.toolbarButton} ${pageType !== 'mobile' ? classes.largeButton : ''}`}
      classes={{ label: classes.buttonLabel }}
      onClick={() => toggleDrawerMenu()}
    >
      {drawerOpen ? (
        <>
          <Close />
          <Typography color="inherit" variant="body2">
            <FormattedMessage id="general.close" />
          </Typography>
        </>
      ) : (
        <>
          <Menu />
          <Typography color="inherit" variant="body2">
            <FormattedMessage id="general.menu" />
          </Typography>
        </>
      )}
    </Button>
  );
};

MenuButton.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  drawerOpen: PropTypes.bool.isRequired,
  pageType: PropTypes.string,
  toggleDrawerMenu: PropTypes.func.isRequired,
};

MenuButton.defaultProps = {
  pageType: null,
};

export default MenuButton;
