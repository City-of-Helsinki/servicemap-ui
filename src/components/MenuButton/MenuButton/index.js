import { withStyles } from '@mui/styles';
import { injectIntl } from 'react-intl';
import MenuButton from './MenuButton';
import styles from '../styles';

export default injectIntl(withStyles(styles)(MenuButton));
