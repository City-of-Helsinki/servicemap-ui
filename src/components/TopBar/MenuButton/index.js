import { withStyles } from '@mui/styles';
import MenuButton from './MenuButton';
import styles from '../styles';

export default withStyles(styles)(MenuButton);
