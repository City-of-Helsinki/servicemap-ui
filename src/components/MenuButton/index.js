import DropDownMenuButton from './DropDownMenuButton';
import OwnSettingsMenuButton from './OwnSettingsMenuButton';

export {
  DropDownMenuButton,
  OwnSettingsMenuButton,
};
