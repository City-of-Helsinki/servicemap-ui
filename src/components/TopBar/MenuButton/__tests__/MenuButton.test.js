import React from 'react';
import MenuButton from '../MenuButton';
import { getRenderWithProviders } from '../../../../../jestUtils';
import englishTranslations from '../../../../i18n/en';


const mockProps = {
  classes: {},
  drawerOpen: false,
  toggleDrawerMenu: () => {},
};

const renderWithProviders = getRenderWithProviders({});


describe('<MenuButton />', () => {

  it('should work', () => {
    const { container } = renderWithProviders(<MenuButton {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does use correct aria arrtibutes', () => {
    const { getByLabelText } = renderWithProviders(<MenuButton {...mockProps} />);
    /*
      The following aria-attributes are based on the accessibility testing report from 26.4.2021
    */
    // Expect button aria-haspopup value to be true
    expect(getByLabelText(englishTranslations['general.menu']).getAttribute('aria-haspopup')).toEqual('true');
    // Expect button aria-label value to be Valikko
    expect(getByLabelText(englishTranslations['general.menu']).getAttribute('aria-label')).toEqual(englishTranslations['general.menu']);
    // // Expect button aria-expanded value to be same as from props
    expect(getByLabelText(englishTranslations['general.menu']).getAttribute('aria-expanded')).toEqual(`${mockProps.drawerOpen}`);
  });
});
