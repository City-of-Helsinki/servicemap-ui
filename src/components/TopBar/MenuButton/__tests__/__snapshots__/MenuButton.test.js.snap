// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuButton /> should work 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Menu"
    class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root undefined undefined css-1p2nkht-MuiButtonBase-root-MuiButton-root"
    id="MenuButton"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6yqi1c-MuiSvgIcon-root"
      data-testid="MenuIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
      />
    </svg>
    <p
      class="MuiTypography-root MuiTypography-body2 css-q91igz-MuiTypography-root"
    >
      Menu
    </p>
  </button>
</div>
`;
